File: packages/kotlin/ttr-semantics/src/main/kotlin/org/tatrman/ttr/semantics/Resolver.kt
package org.tatrman.ttr.semantics

/**
 * One step in the six-step resolution chain (contract §4.3). Ordered as tried.
 */
enum class ResolutionStep { Lexical, SamePackage, NamedImport, WildcardImport, AutoImport, FullyQualified }

/** A single resolution attempt, recorded for diagnostics (TS `ResolutionAttempt`). */
data class ResolutionAttempt(
    val step: ResolutionStep,
    val candidate: String,
    val reason: String? = null,
)

/**
 * Outcome of resolving one reference. Mirrors the TS `ResolutionResult` union:
 * either [Resolved] with the winning step, or [Unresolved] carrying the reason
 * (not-found vs ambiguous), the attempts tried, and any ambiguous candidates.
 */
sealed interface ResolutionResult {
    data class Resolved(
        val symbol: SymbolEntry,
        val viaStep: ResolutionStep,
    ) : ResolutionResult

    data class Unresolved(
        val reason: Reason,
        val tried: List<ResolutionAttempt>,
        val candidates: List<SymbolEntry> = emptyList(),
    ) : ResolutionResult

    enum class Reason { NotFound, Ambiguous }
}

/**
 * Context for resolving a (possibly dotted) reference. Carries the referring
 * document's schema/namespace/package plus the optional enclosing-def qname and
 * import list. Mirrors TS `ResolutionContext`.
 */
data class ResolutionContext(
    val schemaCode: String,
    val namespace: String,
    val enclosingQname: String? = null,
    val imports: List<org.tatrman.ttr.parser.model.ImportStatement>? = null,
    val packageName: String? = null,
)

/** Lexical scope for bare-id resolution (TS `LexicalScope`). */
data class LexicalScope(
    val schemaCode: String,
    val namespace: String,
    val enclosing: Enclosing? = null,
) {
    data class Enclosing(
        val kind: String,
        val qname: String,
    )
}

/**
 * Six-step reference resolver. Mirrors `packages/semantics/src/resolver.ts`
 * exactly: lexical → same-package → named-import → wildcard-import (non-recursive)
 * → cnc auto-import → fully-qualified-but-unique. First hit wins.
 */
class Resolver(
    private val symbols: SymbolTable,
) {
    data class Ref(
        val path: String,
        val parts: List<String>,
    )

    fun resolveReference(
        ref: Ref,
        context: ResolutionContext,
    ): ResolutionResult {
        val tried = mutableListOf<ResolutionAttempt>()

        // Step 1 — lexical (child of the enclosing def).
        val enclosingCandidate = context.enclosingQname?.let { "$it.${ref.path}" }
        if (enclosingCandidate != null) {
            tried += ResolutionAttempt(ResolutionStep.Lexical, enclosingCandidate)
            symbols.get(enclosingCandidate)?.let { return ResolutionResult.Resolved(it, ResolutionStep.Lexical) }
            tried.markLastUnknown()
        }

        // Step 2 — same-package (by constructed qname, then by package + name).
        val fullQname = "${context.schemaCode}.${context.namespace}.${ref.path}"
        if (fullQname != context.enclosingQname) {
            tried += ResolutionAttempt(ResolutionStep.SamePackage, fullQname)
            symbols.get(fullQname)?.let { return ResolutionResult.Resolved(it, ResolutionStep.SamePackage) }
            tried.markLastUnknown()
        }
        if (context.packageName != null) {
            for (entry in symbols.getByPackage(context.packageName)) {
                if (entry.name == ref.path && entry.qname != enclosingCandidate && entry.qname != fullQname) {
                    return ResolutionResult.Resolved(entry, ResolutionStep.SamePackage)
                }
            }
        }

        // Steps 3 & 4 — named / wildcard imports.
        val imports = context.imports
        if (imports != null) {
            for (imp in imports) {
                if (imp.wildcard) continue
                if (imp.target.endsWith(".${ref.path}")) {
                    symbols.get(imp.target)?.let {
                        return ResolutionResult.Resolved(it, ResolutionStep.NamedImport)
                    }
                }
            }

            val wildcardMatches = mutableListOf<SymbolEntry>()
            for (imp in imports) {
                if (!imp.wildcard) continue
                val matches = symbols.getByPackage(imp.target).filter { it.name == ref.path && it.qname != fullQname }
                for (m in matches) {
                    if (wildcardMatches.none { it.qname == m.qname }) wildcardMatches += m
                }
            }
            if (wildcardMatches.size > 1) {
                for (m in wildcardMatches) {
                    tried += ResolutionAttempt(ResolutionStep.WildcardImport, m.qname, "ambiguous")
                }
                return ResolutionResult.Unresolved(ResolutionResult.Reason.Ambiguous, tried, wildcardMatches)
            }
            if (wildcardMatches.size == 1) {
                tried += ResolutionAttempt(ResolutionStep.WildcardImport, wildcardMatches[0].qname)
                return ResolutionResult.Resolved(wildcardMatches[0], ResolutionStep.WildcardImport)
            }
        }

        // Step 5 — cnc.* auto-import (stock roles keyed cnc.role.*, no doubling — D15).
        val cncQname = "cnc.role.${ref.path}"
        if (cncQname != fullQname && cncQname != context.enclosingQname) {
            tried += ResolutionAttempt(ResolutionStep.AutoImport, cncQname)
            symbols.get(cncQname)?.let { return ResolutionResult.Resolved(it, ResolutionStep.AutoImport) }
            tried.markLastUnknown()
        }

        // Step 6 — fully-qualified-but-unique across the project. Try the written
        // form, then the classified tail — the trailing name path with leading
        // model/package/schema/kind segments stripped — so a reference naming a
        // slot the v4.0 key now carries explicitly (`db.dbo.Orders` →
        // `…db.dbo.table.Orders`) still resolves by suffix.
        val tail = classifiedTail(ref.path)
        val forms = if (tail.isNotEmpty() && tail != ref.path) listOf(ref.path, tail) else listOf(ref.path)
        for (cand in forms) {
            val uniqueMatches =
                symbols
                    .getBySuffix(cand)
                    .filter { it.qname != fullQname && it.qname != cncQname }
                    .distinctBy { it.qname }
            if (uniqueMatches.size == 1) {
                tried += ResolutionAttempt(ResolutionStep.FullyQualified, uniqueMatches[0].qname)
                return ResolutionResult.Resolved(uniqueMatches[0], ResolutionStep.FullyQualified)
            }
        }

        return ResolutionResult.Unresolved(ResolutionResult.Reason.NotFound, tried)
    }

    /** Kind segments that may lead a written reference (camelCase + MD aliases). */
    private val kindKeywords =
        setOf(
            "entity",
            "attribute",
            "relation",
            "table",
            "view",
            "column",
            "index",
            "constraint",
            "fk",
            "procedure",
            "query",
            "drillMap",
            "role",
            "project",
            "area",
            "er2dbEntity",
            "er2dbAttribute",
            "er2dbRelation",
            "er2cncRole",
            "domain",
            "map",
            "dimension",
            "hierarchy",
            "measure",
            "cubelet",
            "md2db_cubelet",
            "md2db_domain",
            "md2db_map",
            "md2er_cubelet",
        )

    /**
     * The trailing name path of a dotted reference, with any leading
     * model/package/schema/kind segments stripped (D4 slot order). Mirrors TS
     * `Resolver.classifiedTail` / `classifyReference`.
     */
    private fun classifiedTail(path: String): String {
        val segs = path.split('.').filter { it.isNotEmpty() }
        val packages = symbols.listPackages().filter { it.isNotEmpty() }.toSet()
        val schemas = mutableSetOf("dbo")
        for (q in symbols.allQnames()) {
            val parts = q.split('.')
            val i = parts.indexOf("db")
            if (i >= 0 && i + 1 < parts.size) schemas += parts[i + 1]
        }
        var i = 0
        if (i < segs.size && segs[i] in MODEL_CODES) i++
        run {
            var best = -1
            for (j in i until segs.size) {
                if (segs.subList(i, j + 1).joinToString(".") in packages) best = j
            }
            if (best >= i) i = best + 1
        }
        if (i < segs.size && segs[i] in schemas) i++
        if (i < segs.size && segs[i] in MODEL_CODES) i++
        // schema may follow the post-package model (`pkg.db.dbo.x`).
        if (i < segs.size && segs[i] in schemas) i++
        if (i < segs.size && segs[i] in kindKeywords) i++
        return segs.subList(i, segs.size).joinToString(".")
    }

    fun resolveBareId(
        name: String,
        scope: LexicalScope,
    ): ResolutionResult {
        val tried = mutableListOf<ResolutionAttempt>()

        if (scope.enclosing != null) {
            val withEnclosing = "${scope.enclosing.qname}.$name"
            tried += ResolutionAttempt(ResolutionStep.Lexical, withEnclosing)
            symbols.get(withEnclosing)?.let { return ResolutionResult.Resolved(it, ResolutionStep.Lexical) }
            tried.markLastUnknown()
        }

        val withSchema = "${scope.schemaCode}.${scope.namespace}.$name"
        if (withSchema != scope.enclosing?.qname) {
            tried += ResolutionAttempt(ResolutionStep.SamePackage, withSchema)
            symbols.get(withSchema)?.let { return ResolutionResult.Resolved(it, ResolutionStep.SamePackage) }
            tried.markLastUnknown()
        }

        val cncQname = "cnc.role.$name"
        if (cncQname != withSchema) {
            tried += ResolutionAttempt(ResolutionStep.AutoImport, cncQname)
            symbols.get(cncQname)?.let { return ResolutionResult.Resolved(it, ResolutionStep.AutoImport) }
            tried.markLastUnknown()
        }

        return ResolutionResult.Unresolved(ResolutionResult.Reason.NotFound, tried)
    }
}

/** Stamp `reason = "unknown-symbol"` on the most recent attempt (mirrors TS). */
private fun MutableList<ResolutionAttempt>.markLastUnknown() {
    if (isNotEmpty()) this[lastIndex] = this[lastIndex].copy(reason = "unknown-symbol")
}

File: packages/python/ttr-parser/src/ttr_parser/semantics/resolver.py
"""Reference resolver — the six-step chain (← `resolver.ts`, contracts §3.3).

The step order is load-bearing and pinned by the §5.1 conformance dump:

  1. lexical          — bare id as a child of the enclosing def
  2. same-package     — `schema.ns.name` direct hit, then a sibling in the
                        document's package
  3. named-import     — an `import a.b.c.name` whose suffix matches
  4. wildcard-import  — `import a.b.*`, NON-recursive (exactly one segment below
                        the imported package); two matches ⇒ ambiguous
  5. auto-import      — stock cnc roles under the doubled `cnc.cnc.role.<name>`
  6. fully-qualified  — a unique suffix match anywhere in the project

A named import shadows a wildcard (step 3 precedes step 4). Each failed step is
recorded in `tried` for the not-found diagnostic. Mirrors `resolver.ts` exactly;
the Python API takes the reference as a dotted string (split internally) rather
than the TS `{ path, parts }`.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from ..model import ImportStatement
from .default_schema import MODEL_CODES
from .symbol_table import SymbolEntry, SymbolTable

# Kind segments that may lead a written reference (camelCase + MD aliases).
_KIND_KEYWORDS: frozenset[str] = frozenset(
    {
        "entity", "attribute", "relation", "table", "view", "column", "index",
        "constraint", "fk", "procedure", "query", "drillMap", "role", "project",
        "area", "er2dbEntity", "er2dbAttribute", "er2dbRelation", "er2cncRole",
        "domain", "map", "dimension", "hierarchy", "measure", "cubelet",
        "md2db_cubelet", "md2db_domain", "md2db_map", "md2er_cubelet",
    }
)

ResolutionStep = Literal[
    "lexical",
    "same-package",
    "named-import",
    "wildcard-import",
    "auto-import",
    "fully-qualified",
]

ResolutionReason = Literal[
    "unknown-symbol",
    "not-imported",
    "wildcard-non-recursive",
    "shadowed-by-named-import",
    "lexical-scope-empty",
    "ambiguous",
]


@dataclass(frozen=True, slots=True)
class ResolutionContext:
    schema_code: str
    namespace: str
    imports: tuple[ImportStatement, ...] = ()
    package_name: str = ""
    enclosing_qname: str | None = None


@dataclass(frozen=True, slots=True)
class EnclosingDef:
    kind: str
    qname: str


@dataclass(frozen=True, slots=True)
class LexicalScope:
    schema_code: str
    namespace: str
    enclosing: EnclosingDef | None = None


@dataclass(frozen=True, slots=True)
class ResolutionAttempt:
    step: ResolutionStep
    candidate: str
    reason: ResolutionReason | None = None


@dataclass(frozen=True, slots=True)
class Resolved:
    symbol: SymbolEntry
    via_step: ResolutionStep


@dataclass(frozen=True, slots=True)
class Unresolved:
    reason: Literal["not-found", "ambiguous"]
    tried: tuple[ResolutionAttempt, ...]
    candidates: tuple[SymbolEntry, ...] = ()


ResolutionResult = Resolved | Unresolved


class Resolver:
    def __init__(self, symbols: SymbolTable) -> None:
        self._symbols = symbols

    def get_symbol(self, qname: str) -> SymbolEntry | None:
        """Direct symbol-table lookup by fully-qualified name."""
        return self._symbols.get(qname)

    def resolve_reference(
        self, ref: str, context: ResolutionContext
    ) -> ResolutionResult:
        path = ref
        tried: list[ResolutionAttempt] = []

        enclosing_candidate = (
            f"{context.enclosing_qname}.{path}" if context.enclosing_qname else None
        )

        # Step 1 — lexical scope.
        if enclosing_candidate is not None:
            symbol = self._symbols.get(enclosing_candidate)
            if symbol is not None:
                return Resolved(symbol, "lexical")
            tried.append(
                ResolutionAttempt("lexical", enclosing_candidate, "unknown-symbol")
            )

        # Step 2 — same-package: direct `schema.ns.name`, then package sibling.
        full_qname = f"{context.schema_code}.{context.namespace}.{path}"
        if full_qname != context.enclosing_qname:
            symbol = self._symbols.get(full_qname)
            if symbol is not None:
                return Resolved(symbol, "same-package")
            tried.append(
                ResolutionAttempt("same-package", full_qname, "unknown-symbol")
            )

        if context.package_name:
            for entry in self._symbols.get_by_package(context.package_name):
                if entry.name == path:
                    candidate = entry.qname.value
                    if candidate not in (enclosing_candidate, full_qname):
                        return Resolved(entry, "same-package")

        # Steps 3–4 — imports.
        if context.imports:
            for imp in context.imports:
                if imp.wildcard:
                    continue
                if imp.target.endswith(f".{path}"):
                    symbol = self._symbols.get(imp.target)
                    if symbol is not None:
                        return Resolved(symbol, "named-import")

            wildcard_matches: list[SymbolEntry] = []
            for imp in context.imports:
                if not imp.wildcard:
                    continue
                for entry in self._symbols.get_by_package(imp.target):
                    if entry.name == path and entry.qname.value != full_qname:
                        if not any(
                            w.qname.value == entry.qname.value for w in wildcard_matches
                        ):
                            wildcard_matches.append(entry)

            if len(wildcard_matches) > 1:
                for match in wildcard_matches:
                    tried.append(
                        ResolutionAttempt(
                            "wildcard-import", match.qname.value, "ambiguous"
                        )
                    )
                return Unresolved("ambiguous", tuple(tried), tuple(wildcard_matches))
            if len(wildcard_matches) == 1:
                tried.append(
                    ResolutionAttempt("wildcard-import", wildcard_matches[0].qname.value)
                )
                return Resolved(wildcard_matches[0], "wildcard-import")

        # Step 5 — auto-import (stock cnc roles, keyed cnc.role.* — D15).
        cnc_qname = f"cnc.role.{path}"
        if cnc_qname not in (full_qname, context.enclosing_qname):
            symbol = self._symbols.get(cnc_qname)
            if symbol is not None:
                return Resolved(symbol, "auto-import")
            tried.append(ResolutionAttempt("auto-import", cnc_qname, "unknown-symbol"))

        # Step 6 — fully-qualified: the written form, then the classified tail —
        # the trailing name path with leading model/package/schema/kind segments
        # stripped — so a reference naming a slot the v4.0 key now carries
        # explicitly (`db.dbo.Orders` → `…db.dbo.table.Orders`) resolves by suffix.
        tail = self._classified_tail(path)
        forms = [path, tail] if tail and tail != path else [path]
        for cand in forms:
            seen: set[str] = set()
            unique_matches = []
            for e in self._symbols.get_by_suffix(cand):
                if e.qname.value in (full_qname, cnc_qname) or e.qname.value in seen:
                    continue
                seen.add(e.qname.value)
                unique_matches.append(e)
            if len(unique_matches) == 1:
                tried.append(
                    ResolutionAttempt("fully-qualified", unique_matches[0].qname.value)
                )
                return Resolved(unique_matches[0], "fully-qualified")

        return Unresolved("not-found", tuple(tried))

    def _classified_tail(self, path: str) -> str:
        """The trailing name path with leading model/package/schema/kind segments
        stripped (← TS ``Resolver.classifiedTail`` / ``classifyReference``)."""
        segs = [s for s in path.split(".") if s]
        all_entries = self._symbols.get_all()
        packages = {e.package_name for e in all_entries if e.package_name}
        schemas = {"dbo"}
        for e in all_entries:
            parts = e.qname.value.split(".")
            if "db" in parts:
                k = parts.index("db")
                if k + 1 < len(parts):
                    schemas.add(parts[k + 1])
        i = 0
        if i < len(segs) and segs[i] in MODEL_CODES:
            i += 1
        best = -1
        for j in range(i, len(segs)):
            if ".".join(segs[i : j + 1]) in packages:
                best = j
        if best >= i:
            i = best + 1
        if i < len(segs) and segs[i] in schemas:
            i += 1
        if i < len(segs) and segs[i] in MODEL_CODES:
            i += 1
        # schema may follow the post-package model (`pkg.db.dbo.x`).
        if i < len(segs) and segs[i] in schemas:
            i += 1
        if i < len(segs) and segs[i] in _KIND_KEYWORDS:
            i += 1
        return ".".join(segs[i:])

    def resolve_bare_id(self, name: str, scope: LexicalScope) -> ResolutionResult:
        tried: list[ResolutionAttempt] = []

        if scope.enclosing is not None:
            with_enclosing = f"{scope.enclosing.qname}.{name}"
            symbol = self._symbols.get(with_enclosing)
            if symbol is not None:
                return Resolved(symbol, "lexical")
            tried.append(
                ResolutionAttempt("lexical", with_enclosing, "unknown-symbol")
            )

        with_schema = f"{scope.schema_code}.{scope.namespace}.{name}"
        enclosing_qname = scope.enclosing.qname if scope.enclosing else None
        if with_schema != enclosing_qname:
            symbol = self._symbols.get(with_schema)
            if symbol is not None:
                return Resolved(symbol, "same-package")
            tried.append(
                ResolutionAttempt("same-package", with_schema, "unknown-symbol")
            )

        cnc_qname = f"cnc.role.{name}"
        if cnc_qname != with_schema:
            symbol = self._symbols.get(cnc_qname)
            if symbol is not None:
                return Resolved(symbol, "auto-import")
            tried.append(ResolutionAttempt("auto-import", cnc_qname, "unknown-symbol"))

        return Unresolved("not-found", tuple(tried))
